precompiles/common/precompile.go

```diff
@@ -53,20 +53,65 @@ func (p Precompile) RunSetup(
 
 	// NOTE: This is a special case where the calling transaction does not specify a function name.
 	// In this case we default to a `fallback` or `receive` function on the contract.
-	if len(contract.Input) != 0 {
+
+	// Check if the fallback or receive functions are present in the ABI
+	isFallbackPresent := p.HasFallback()
+	isReceivePresent := p.HasReceive()
+
+	// Simplify the calldata checks
+	isEmptyCallData := len(contract.Input) == 0
+	isShortCallData := len(contract.Input) < 4
+	isStandardCallData := len(contract.Input) >= 4
+
+	// Case 1: Calldata is empty
+	if isEmptyCallData {
+		switch {
+		// Case 1.1: Send call or transfer tx - 'receive' is called if present and value is transferred
+		case contract.Value().Sign() > 0 && isReceivePresent:
+			method = &p.Receive
+		// Case 1.2: Either 'receive' is not present, or no value is transferred - call 'fallback' if present
+		case isFallbackPresent:
+			method = &p.Fallback
+		// Case 1.3: Neither 'receive' nor 'fallback' are present - return error
+		default:
+			return sdk.Context{}, nil, nil, uint64(0), nil, vm.ErrExecutionReverted
+		}
+	}
+
+	// Case 2: calldata is non-empty but less than 4 bytes needed for a method
+	// Case 2.1: calldata contains less than 4 bytes needed for a method - 'fallback' is called if present
+	if isShortCallData && isFallbackPresent {
```

Comment:
if any of the above cases is executed (eg. 1.2) then this `if` will still be checked, we should refactor it to a `switch`